### PR TITLE
add KSP processor and implement it for compose

### PR DESCRIPTION
**Pull request by gabrielittner**
First half of the KSP implementation, the fragment based compose and renderer annotations will be handled in the next PR.

### CODE REVIEW

File: codegen-compiler/src/main/kotlin/com/freeletics/khonshu/codegen/KhonshuSymbolProcessor.kt [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+    private inline fun <reified T : KSAnnotated, reified A : Annotation> Resolver.generateCodeForAnnotation(
+        parser: T.(A) -> BaseData?,
+    ) {
+        getSymbolsWithAnnotation(A::class.qualifiedName!!)
+            .forEach {
+                if (it !is T) {
+                    val target = when (T::class) {
+                        KSFunctionDeclaration::class -> "functions"
+                        KSClassDeclaration::class -> "classes"
```
Comment (gabrielittner) [inline codegen-compiler/src/main/kotlin/com/freeletics/khonshu/codegen/KhonshuSymbolProcessor.kt:54]: Supporting `KSClassDeclaration` here is already for annotations on renderers

File: codegen-compiler/src/main/kotlin/com/freeletics/khonshu/codegen/parser/ksp/CommonParser.kt [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        var superTypeName = superType.toTypeName(parentTypeParameters.toTypeParameterResolver())
+        if (parentTypeArguments.isNotEmpty() && superTypeName is ParameterizedTypeName) {
+            superTypeName = superTypeName.updateWith(parentTypeArguments, parentTypeParameters)
+        }
+        yield(superTypeName)
+
+        val superDeclaration = superType.declaration as? KSClassDeclaration ?: return@forEach
+        yieldAll(superDeclaration.allSuperTypes(superTypeName))
+    }
+}
```
Comment (gabrielittner) [inline codegen-compiler/src/main/kotlin/com/freeletics/khonshu/codegen/parser/ksp/CommonParser.kt:99]: This and the method below is the KSP equivalent of what I changed in #493

File: codegen-compiler/src/test/kotlin/com/freeletics/khonshu/codegen/parser/ksp/AllSuperTypesTest.kt [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+import com.squareup.kotlinpoet.INT
+import com.squareup.kotlinpoet.LONG
+import com.squareup.kotlinpoet.ParameterizedTypeName
+import com.squareup.kotlinpoet.SHORT
+import com.squareup.kotlinpoet.STRING
+import com.squareup.kotlinpoet.TypeName
+import com.tschuchort.compiletesting.KotlinCompilation.ExitCode
+import org.junit.Test
+
+internal class AllSuperTypesTest {
```
Comment (gabrielittner) [inline codegen-compiler/src/test/kotlin/com/freeletics/khonshu/codegen/parser/ksp/AllSuperTypesTest.kt:19]: Copy of the existing Anvil test for this

**Review by kboyarshinov — APPROVED** [review]
LGTM 👍

**Pull request merged**